p2p/dandelion/src/txpool.rs

```diff
@@ -0,0 +1,545 @@
+
+                }
+            }
+        }
+    }
+}
+
+/// The dandelion pool service builder.
+pub struct DandelionPoolServiceBuilder<P, R, Tx, TxID, PID> {
+    /// The dandelion++ router.
+    router: Option<R>,
+    /// The backing transaction pool.
+    backing_pool: Option<P>,
+    /// The d++ config.
+    config: Option<DandelionConfig>,
+
+    _types: PhantomData<(Tx, TxID, PID)>,
+}
+
+impl<P, R, Tx, TxID, PID> Default for DandelionPoolServiceBuilder<P, R, Tx, TxID, PID> {
+    fn default() -> Self {
+        Self {
+            router: None,
+            backing_pool: None,
+            config: None,
+
+            _types: PhantomData,
+        }
+    }
+}
+
+impl<P, R, Tx, TxID, PID> DandelionPoolServiceBuilder<P, R, Tx, TxID, PID>
+where
+    Tx: Clone + Send + 'static,
+    TxID: Hash + Eq + Clone + Send + 'static,
+    PID: Hash + Eq + Clone + Send + 'static,
+    P: Service<
+            TxStoreRequest<Tx, TxID>,
+            Response = TxStoreResponse<Tx, TxID>,
+            Error = tower::BoxError,
+        > + Send
+        + 'static,
+    P::Future: Send + 'static,
+    R: Service<DandelionRouteReq<Tx, PID>, Response = State, Error = DandelionRouterError>
+        + Send
+        + 'static,
+    R::Future: Send + 'static,
+{
+    /// Starts a new builder.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Adds the router to the builder.
+    pub fn with_router(self, router: R) -> Self {
+        Self {
+            router: Some(router),
+            ..self
+        }
+    }
+
+    /// Adds the backing pool to the builder.
+    pub fn with_backing_pool(self, backing_pool: P) -> Self {
+        Self {
+            backing_pool: Some(backing_pool),
+            ..self
+        }
+    }
+
+    /// Adds the config to the builder.
+    pub fn with_config(self, config: DandelionConfig) -> Self {
+        Self {
+            config: Some(config),
+            ..self
+        }
+    }
+
+    pub fn spawn(self, buffer: usize) -> DandelionPoolService<Tx, TxID, PID> {
+        let config = self.config.expect("Config was not added to builder");
+
+        let (tx, rx) = mpsc::channel(buffer);
+
+        let pool = DandelionPool {
+            dandelion_router: self
+                .router
+                .expect("Dandelion router was not added to builder."),
+            backing_pool: self
+                .backing_pool
+                .expect("Backing pool was not added to router"),
+            routing_set: JoinSet::new(),
+            stem_origins: HashMap::new(),
+            embargo_timers: DelayQueue::new(),
+            embargo_dist: Exp::new(1.0 / config.average_embargo_timeout().as_secs_f64()).unwrap(),
+            config,
+            _tx: PhantomData,
+        };
+
+        let span = tracing::debug_span!("dandelion_pool");
+
+        tokio::spawn(pool.run(rx).instrument(span));
+
+        DandelionPoolService {
+            tx: PollSender::new(tx),
+        }
+    }
+}
```

Comment:
Either no builder or panic docs please, there's no mention that fields must be filled on this one.